File: Data Structures & Algorithms/partition-labels/submission-10.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        end = {}

        for i, c in enumerate(s): 
            end[c] = i  

        res = [] 

        l = 0 
        curMax = 0
        for r in range(len(s)): 
            curMax = max(curMax, end[s[r]])
            if r == curMax: 
               res.append(r - l + 1)  
               l = r + 1 

        return res 

File: Data Structures & Algorithms/eating-bananas/submission-10.js
class Solution {
    /**
     * @param {number[]} piles
     * @param {number} h
     * @return {number}
     */
    minEatingSpeed(piles, h) { 
        let l = 1 
        let r = Math.max(...piles)  
        let res = r

        while(l <= r) {
            let mid = Math.floor((r + l) / 2) 
            
            let need = 0 
            for (const pile of piles) {
                need += Math.ceil(pile / mid)
            } 

            if (need <= h) {
                res = Math.min(mid, res) 
                r = mid - 1
            } else {
                l = mid + 1
            }
        } 

        return res
    }
}
